src/Clapi/Valuespace.hs

```diff
@@ -305,31 +306,39 @@ rectifyTypes vs explicitPaths =
     populateDefsFor deflessNodes vs = if null deflessNodes then (mempty, vs) else let
         np = head $ Set.toList deflessNodes
       in do
-        vs' <- case getOrBuildDef deflessNodes np of
-            Left err -> (Map.fromList [(np, err)], vs)
-            Right (tp, md) ->
-                ( mempty
-                , over types (Mos.setDependency np tp) $ over defs (Map.insert tp md) vs)
-        populateDefsFor (Set.delete np deflessNodes) vs'
-
-    getOrInferType :: Set.Set NodePath -> NodePath -> CanFail TypePath
-    getOrInferType dirtyPaths np = if Set.member np dirtyPaths
-      then maybe (parentDerivedType dirtyPaths np) Right $ Map.lookup np explicitlyAssigned
-      else getType vs np
-    parentDerivedType :: Set.Set NodePath -> NodePath -> CanFail TypePath
-    parentDerivedType dirtyPaths np = case splitBasename np of
-        Just (pp, cn) ->
-            getOrBuildDef dirtyPaths pp >>=
-            return . snd >>=
-            note "Parent has no type for child" . flip childTypeFor cn
-        Nothing -> error "Hit root deriving parent types, code bad."
-    getOrBuildDef :: Set.Set NodePath -> NodePath -> CanFail (TypePath, Definition)
-    getOrBuildDef dirtyPaths np = do
-        tp <- getOrInferType dirtyPaths np
-        mt <- getOrInferType dirtyPaths tp >>= categoriseMetaTypePath
-        tn <- getNode tp vs
-        d <- validateTypeNode mt tn
-        return (tp, d)
+        case runState (runExceptT (getOrBuildDef np)) (deflessNodes, vs) of
+            (Left err, (deflessNodes', vs')) -> (Map.fromList [(np, err)], vs)
+            (Right _, (deflessNodes', vs')) -> populateDefsFor deflessNodes' vs'
+
+    getOrInferType :: NodePath -> ExceptT String (State (Set.Set NodePath, Valuespace v)) TypePath
+    getOrInferType np = do
+        (dirtyPaths, vs) <- get
+        if Set.member np dirtyPaths
+            then maybe (parentDerivedType np) return (Map.lookup np explicitlyAssigned)
+            else cfet $ getType vs np
+    parentDerivedType :: NodePath -> ExceptT String (State (Set.Set NodePath, Valuespace v)) TypePath
+    parentDerivedType np = case splitBasename np of
+        Just (pp, cn) -> getOrBuildDef pp >>= cfet . note "Parent has no child for type" . flip childTypeFor cn
+        Nothing -> return $ error "Hit root deriving parent types, code bad."
+    getOrBuildDef :: NodePath -> ExceptT String (State (Set.Set NodePath, Valuespace v)) Definition
+    getOrBuildDef np = do
+        tp <- getOrInferType np
+        mtp <- getOrInferType tp
+        mt <- cfet $ categoriseMetaTypePath mtp
+        tn <- cfet $ getNode tp vs
+        md <- cfet $ validateTypeNode mt tn
+        modify $ \(dirtyPaths, vs) -> let
```

Comment:
This `vs` shadows the one from the parent definition. I was getting confused with how you got hold of it on 328 before you'd extracted it from the state, until I realised they were different! I presume you're safe to use the other `vs` on 328 because you're never mutating the data content of the tree, only the type cache?

Reply:
It is safe, but as you say unclear. I'm trying to get this function to use the cache for clean ones (so it reads the state earlier) and I'm going to unshadow for clarity as part of that (if I can get it working).